### PR TITLE
Remove old draft assets from Asset Manager when they are no longer used

### PR DESCRIPTION
**Pull request by kevindew**
Trello: https://trello.com/c/5JMGsuBj/581-follow-up-work-from-versioning-review

This uses the point of creating a preview as a point to delete assets
from Asset Manager which are no longer used.

When an edition is sent to the preview service the previous revision
assets are compared with the current revision and any that are in a
draft state are then deleted from Asset Manager.

This doesn't feel like the most robust of solutions as if assets are not
removed at this point they will remain in draft forever. The approach
that would resolve that would be to have some form of scheduled task
that performs a cleanup of assets periodically. However since we don't
have prior art currently in scheduled tasks on this app I have not
implemented it.

### CODE REVIEW

**Comment by kevindew** [comment]
@benthorner following on from our IRL chat on Friday.

Like discussed, a scheduled task is out of the scope from this card. May well prioritise once we have some scheduling art off the back of scheduling work.

I've put together the DraftAssetCleanupService as discussed and have slimmed down the tests for `create_preview` and `try_create_preview`. There's still a bit of repetition since the try_create_preview is close to being `create_preview` with a boolean (although not quite because of the `has_issues?` inconsistency)

I also moved around the exception catching to be done at method level.

File: spec/services/preview_service_spec.rb [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        expect { PreviewService.new(edition).try_create_preview }
+          .not_to raise_error
+        expect(edition.revision_synced).to be(false)
+      end
+    end
+
+    context "when there are pre-preview issues" do
+      let(:edition) { create(:edition, title: "", revision_synced: true) }
+
+      it "marks the edition as revision not synced and doesn't sent to the Publishing API" do
```
Comment (benthorner) [inline spec/services/preview_service_spec.rb]: I think this would be a bit easier to read it the 'and's were separate 'it's. We seem to do that pretty consistently everywhere else. What do you think?

File: spec/services/preview_service_spec.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      expect(request).to have_been_requested
+    end
+
+    it "marks the edition as revision synced" do
+      edition = create(:edition, revision_synced: false)
+      PreviewService.new(edition).create_preview
+      expect(edition.reload.revision_synced).to be(true)
+    end
+
+    it "cleans up draft assets" do
```
Comment (benthorner) [inline spec/services/preview_service_spec.rb]: There's a slight inconsistency between this and the 'try_create_preview' description. How about 'delegates cleaning up draft assets', just to make the content of the test more quickly identifiable?

File: spec/services/preview_service_spec.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+  describe "#create_preview" do
+    it "updates the Publishing API" do
+      edition = create(:edition)
+      request = stub_publishing_api_put_content(edition.content_id, {})
+      PreviewService.new(edition).create_preview
+      expect(request).to have_been_requested
+    end
+
+    it "marks the edition as revision synced" do
```
Comment (benthorner) [inline spec/services/preview_service_spec.rb]: Looking at the test for the new service, how about 'revision_synced'?

File: spec/services/preview_service_spec.rb (hunk 1 above)
Comment (kevindew) [reply spec/services/preview_service_spec.rb]: Agreed have amended

File: spec/services/preview_service_spec.rb (hunk 2 above)
Comment (kevindew) [reply spec/services/preview_service_spec.rb]: Sure

File: spec/services/preview_service_spec.rb (hunk 3 above)
Comment (kevindew) [reply spec/services/preview_service_spec.rb]: I'm not quite sure what you mean here, do you mean `it "marks the edition as 'revision_synced'"`?

**Comment by kevindew** [comment]
> However, shouldn't we be stubbing the DraftAssetCleanupService globally in the PreviewService spec, so it remains a unit test? 

I'd suggest not myself. I'd argue that interfacing with that object is part of the unit of this test. It's also not a common approach we've used. 

Personal preference on this is normally to check that the method does what it's expected which allows the tests to not be concerned with the how, but just not to sweat the edge cases as they'll part of the other service's tests. So doing `expect_any_instance_of(DraftAssetCleanupService).to receive(:call)` feels a bit funky for me - but the setup requirements of avoiding that are pretty high, nor does it feel like a dependency we'd want to inject (where a stub would feel more appropriate).

> Related to that: the asset manager down case is now split in two: between upload_assets and a stubbed exception raised by this new service. I suppose it depends on the testing approach, though...

Yup, either could cause it - we would have same coverage though if that was changed. Given the wording of asset manager down it feels fine to be either case.

If we were handling it different I imagine it'd make sense to have separate whens. Would quite possibly make sense for DraftAssetCleanupService to raise an exception of it's own.

**Comment by kevindew** [comment]
@benthorner lets chat face-to-face once you're back.

**Comment by kevindew** [comment]
@benthorner sorry - just done those two small things.
